util/internal/test/common/src/test/java/org/hibernate/search/util/impl/test/reflect/Experiments2Test.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.util.impl.test.reflect;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.hibernate.search.util.impl.test.function.ThrowingConsumer;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DynamicContainer;
+import org.junit.jupiter.api.DynamicNode;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestFactory;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.DynamicTestInvocationContext;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.InvocationInterceptor;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.platform.commons.util.AnnotationUtils;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+class Experiments2Test {
```

Comment:
This class contains a different approach to this "problem". Here, dynamic tests are used through `@TestFactory`.  With it we can produce a better tree structure for the executed tests:
```
[root]
	testsFromStaticClass()
		Configuration group: [string1] [1] [true]
			test1()
			test2()
		Configuration group: [string2] [2] [true]
			test1()
			test2()
		Configuration group: [string3] [3] [true]
			test1()
			test2()
		Configuration group: [string3] [3] [false]
			test1()
			test2()
	testsInnerClass()
		Configuration group: [string1] [1] [true]
			test1()
			test2()
			test3()
		Configuration group: [string2] [2] [true]
			test1()
			test2()
			test3()
		Configuration group: [string3] [3] [true]
			test1()
			test2()
			test3()
		Configuration group: [string3] [3] [false]
			test1()
			test2()
			test3()
```
The downside is that we can't really add extensions within the "test inner classes", and if we need any extensions, they must live at the root class level.
BeforeEach/AfterEach methods in the root will be treated as BeforeAll/AfterAll for those inner test classes.
If we'd need to execute something around the inner test methods from the root extension, we can try to do so through a `InvocationInterceptor#interceptDynamicTest` and do something before/after executing ` invocation.proceed();`

It's probably easier to run this test in the IDE see how the integration works, and all the output to get better picture 😃 (Note in the IDE you wouldn't be able to jump to the test method from the test panel, since the test.. are dynamic 😃)

Reply:
ohh and we'll need to update jqassistant rules:
```
[ERROR] --[ Constraint Violation ]-----------------------------------------
[ERROR] Constraint: hsearch:NestedTestTypesShouldBeAnnotatedWithNested
[ERROR] Severity: MAJOR
[ERROR] Number of rows: 2
[ERROR]             Nested test types should be annotated with @Nested in order to be actually executed
[ERROR]             when their surrounding type is executed.
[ERROR]             Exceptions are only allowed for:
[ERROR]             - Abstract test classes.
[ERROR]         
[ERROR]   Test=org.hibernate.search.util.impl.test.reflect.Experiments2Test$TheseAreActualTestsButTheyNeedParametersTest
[ERROR]   Test=org.hibernate.search.util.impl.test.reflect.Experiments2Test$TheseAreActualTestsButTheyNeedParametersAndClassIsStaticTest
[ERROR] -------------------------------------------------------------------
```
Maybe having some "special name pattern" for these types of inner tests so that we could exclude them from queries ....